src/trace/connection.rs

```diff
@@ -0,0 +1,188 @@
+use bson::oid::ObjectId;
+
+use crate::{
+    event::cmap::{
+        CmapEventHandler,
+        ConnectionCheckedInEvent,
+        ConnectionCheckedOutEvent,
+        ConnectionCheckoutFailedEvent,
+        ConnectionCheckoutFailedReason,
+        ConnectionCheckoutStartedEvent,
+        ConnectionClosedEvent,
+        ConnectionClosedReason,
+        ConnectionCreatedEvent,
+        ConnectionReadyEvent,
+        PoolClearedEvent,
+        PoolClosedEvent,
+        PoolCreatedEvent,
+        PoolReadyEvent,
+    },
+    trace::{TracingRepresentation, CONNECTION_TRACING_EVENT_TARGET},
+};
+
+#[derive(Clone)]
+pub(crate) struct ConnectionTracingEventEmitter {
+    topology_id: ObjectId,
+}
+
+impl ConnectionTracingEventEmitter {
+    pub(crate) fn new(topology_id: ObjectId) -> ConnectionTracingEventEmitter {
+        Self { topology_id }
+    }
+}
+
+impl CmapEventHandler for ConnectionTracingEventEmitter {
+    fn handle_pool_created_event(&self, event: PoolCreatedEvent) {
+        let options_ref = event.options.as_ref();
+        tracing::debug!(
```

Comment:
This probably has already been discussed so sorry for retreading ground, but does `tracing` handle the non-structured format for us (e.g. if someone is just using `log`)? Is there a way to configure that output so it looks like the spec defined format?

Reply:
yes, tracing handles the conversion for us, and unfortunately I don't believe it is possible for us to configure the output format at this time. the output looks something like "a=b, c=d, e=f"; it is generated by the `Display` implementation here: https://github.com/tokio-rs/tracing/blob/3258a2bb2ae5f1b3be755bffbaef7ff851ec3b4f/tracing/src/lib.rs#L1162-L1205 

I think in order to customize it we would have to add logic to emit log messages from scratch ourselves. I'm also not sure if/how we could prevent our tracing events from getting converted to log messages if a user used `tracing` with the `log` feature on. (not sure if it is possible to check whether a feature flag in another crate has been enabled?) if not, we might end up giving them two copies of every message, the one generating from the tracing event and the one we emit directly.

on top of those issues, I considered this acceptable because users of the `log` feature of tracing will be used to receiving logged key-value pairs in this format, and also because the strings defined in the spec are SHOULDs, not MUSTs, which was done both because we aren't actually testing string representations and because there were different schools of thought amongst reviewers as to whether a concise a=b, c=d, .. form or a more sentence-like form was preferable.

Reply:
Sounds good, thanks for the explanation!